gtsam/nonlinear/NonlinearFactor.h

```diff
@@ -272,503 +307,306 @@ class GTSAM_EXPORT NoiseModelFactor: public NonlinearFactor {
-/** A convenient base class for creating your own NoiseModelFactor with 3
- * variables.  To derive from this class, implement evaluateError(). */
-template<class VALUE1, class VALUE2, class VALUE3>
-class NoiseModelFactor3: public NoiseModelFactor {
-
-public:
-
-  // typedefs for value types pulled from keys
-  typedef VALUE1 X1;
-  typedef VALUE2 X2;
-  typedef VALUE3 X3;
-
-protected:
-
-  typedef NoiseModelFactor Base;
-  typedef NoiseModelFactor3<VALUE1, VALUE2, VALUE3> This;
-
-public:
-
-  /**
-   * Default Constructor for I/O
-   */
-  NoiseModelFactor3() {}
-
-  /**
-   * Constructor
-   * @param noiseModel shared pointer to noise model
-   * @param j1 key of the first variable
-   * @param j2 key of the second variable
-   * @param j3 key of the third variable
-   */
-  NoiseModelFactor3(const SharedNoiseModel& noiseModel, Key j1, Key j2, Key j3) :
-    Base(noiseModel, cref_list_of<3>(j1)(j2)(j3)) {}
-
-  ~NoiseModelFactor3() override {}
-
-  /** methods to retrieve keys */
-  inline Key key1() const { return keys_[0]; }
-  inline Key key2() const { return keys_[1]; }
-  inline Key key3() const { return keys_[2]; }
-
-  /** Calls the 3-key specific version of evaluateError, which is pure virtual
-   * so must be implemented in the derived class. */
-  Vector unwhitenedError(const Values& x, boost::optional<std::vector<Matrix>&> H = boost::none) const override {
-    if(this->active(x)) {
-      if(H)
-        return evaluateError(x.at<X1>(keys_[0]), x.at<X2>(keys_[1]), x.at<X3>(keys_[2]), (*H)[0], (*H)[1], (*H)[2]);
-      else
-        return evaluateError(x.at<X1>(keys_[0]), x.at<X2>(keys_[1]), x.at<X3>(keys_[2]));
-    } else {
-      return Vector::Zero(this->dim());
-    }
+  template <typename CONTAINER,  // use "dummy" parameter T to delay deduction
+            size_t T = N, typename std::enable_if<(T > 1), bool>::type = true>
+  NoiseModelFactorN(const SharedNoiseModel& noiseModel, CONTAINER keys)
+      : Base(noiseModel, keys) {
+    assert(keys.size() == N);
   }
 
-  /**
-   *  Override this method to finish implementing a trinary factor.
-   *  If any of the optional Matrix reference arguments are specified, it should compute
-   *  both the function evaluation and its derivative(s) in X1 (and/or X2, X3).
-   */
-  virtual Vector
-  evaluateError(const X1&, const X2&, const X3&,
-      boost::optional<Matrix&> H1 = boost::none,
-      boost::optional<Matrix&> H2 = boost::none,
-      boost::optional<Matrix&> H3 = boost::none) const = 0;
+  /// @}
 
-private:
+  ~NoiseModelFactorN() override {}
 
-  /** Serialization function */
-  friend class boost::serialization::access;
-  template<class ARCHIVE>
-  void serialize(ARCHIVE & ar, const unsigned int /*version*/) {
-    ar & boost::serialization::make_nvp("NoiseModelFactor",
-        boost::serialization::base_object<Base>(*this));
+  /** Methods to retrieve keys */
+#define SUB(Old, New) template <int Old = New>  // to delay template deduction
```

Comment:
Indent?

Reply:
I can indent, but clang-format auto un-indents it since it's a macro/preprocessor thing (https://google.github.io/styleguide/cppguide.html#Preprocessor_Directives)